installinator/src/peers.rs

```diff
@@ -0,0 +1,161 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{fmt, net::Ipv6Addr, str::FromStr};
+
+use anyhow::{anyhow, bail, Result};
+use async_trait::async_trait;
+use bytes::{Bytes, BytesMut};
+use futures::{prelude::*, stream::FuturesUnordered};
+use itertools::Itertools;
+
+#[derive(Debug)]
+pub(crate) struct Peers {
+    log: slog::Logger,
+    imp: Box<dyn PeersImpl>,
+}
+
+impl Peers {
+    pub(crate) async fn mock_discover(log: &slog::Logger) -> Result<Peers> {
+        let log = log.new(slog::o!("component" => "Peers"));
+        let imp = Box::new(MockPeers::new(&log)?);
+        Ok(Self { log, imp })
+    }
+
+    pub(crate) async fn fetch_artifact(
+        &self,
+        artifact_id: &ArtifactId,
+    ) -> Result<(Ipv6Addr, Bytes)> {
+        // TODO: do we want a check phase that happens before the download?
+
+        let download_futs: FuturesUnordered<_> = self
+            .imp
+            .peers()
+            .iter()
+            .map(|peer| async {
+                let result =
+                    self.imp.fetch_artifact_from_peer(peer, artifact_id).await;
+                (*peer, result)
+            })
+            .collect();
```

Comment:
One thing we may want to do here is rather than using `FuturesUnordered` which bundles everything up into a single task, use `tokio::spawn` + mpsc and broadcast channels. Which is better depends on the size of the artifacts and whether we want to try downloading sequentially or in parallel, I think.